source/tools/AutomatedTestSuite.h

```diff
@@ -0,0 +1,262 @@
+
+constexpr double kHighLowThreshold = 0.1; // Difference between CPUs to consider them BIG-little.
+constexpr double kMaxUtilization = 0.9; // Maximum load that we will push the CPU to.
+
+/**
+ * Run an automated test, analyze the results and print a report.
+ * 1) Determine what CPUs are available
+ * 2) Run VoiceMark on each CPU to determine big vs little CPUs
+ * 3) Run LatencyMark with a light load, a heavy load, then an alternating load.
+ * 4) Print report with analysis.
+ */
+class AutomatedTestSuite : public ITestHarness {
+
+public:
+    AutomatedTestSuite(AudioSinkBase *audioSink,
+            SynthMarkResult *result,
+            LogTool *logTool = nullptr)
+    : mAudioSink(audioSink)
+    , mResult(result)
+    , mLogTool(logTool)
+    {
+        setNumVoices(kSynthmarkNumVoicesLatency);
+
+        // Constant seed to obtain a fixed pattern of pseudo-random voices
+        // for experiment repeatability and reproducibility
+        //srand(time(NULL)); // "Random" seed
+        srand(0); // Fixed seed
+    }
+
+    virtual ~AutomatedTestSuite() {
+    }
+
+    virtual void setNumVoices(int32_t numVoices) {
+        mNumVoices = numVoices;
+    }
+
+    virtual void setDelayNoteOnSeconds(int32_t delayNotesOn) {
+        mDelayNotesOn = delayNotesOn;
+    }
+
+    virtual const char *getName() {
+        return "Automated Test Suite";
+    }
+
+    virtual int32_t runTest(int32_t sampleRate, int32_t framesPerBurst, int32_t numSeconds) {
+        int32_t err = measureLowHighCpuPerformance(sampleRate, framesPerBurst, 10);
+        if (err) return err;
+
+        // Test both sizes of CPU if needed. BIG or little
+        if (mHaveBigLittle) {
+            err = measureLatency(sampleRate, framesPerBurst, numSeconds,
+                                 mLittleCpu, mVoiceMarkMaxLittle);
+            if (err) return err;
+        }
+        err = measureLatency(sampleRate, framesPerBurst, numSeconds,
+                             mBigCpu, mVoiceMarkMaxBig);
+        return err;
+    }
+
+private:
+
+    virtual int32_t measureLowHighCpuPerformance(int32_t sampleRate,
+                                                 int32_t framesPerBurst,
+                                                 int32_t numSeconds) {
+        std::stringstream resultMessage;
+        int numCPUs = HostTools::getCpuCount();
+        SynthMarkResult result1;
+        VoiceMarkHarness *harness = new VoiceMarkHarness(mAudioSink, &result1);
+        harness->setTargetCpuLoad(kMaxUtilization);
+        harness->setInitialVoiceCount(mNumVoices);
+        harness->setDelayNoteOnSeconds(mDelayNotesOn);
+
+        int lowCpu = (1 * numCPUs) / 4;
+        mAudioSink->setRequestedCpu(lowCpu);
+        int32_t err = harness->runTest(sampleRate, framesPerBurst, numSeconds);
+        if (err) return err;
+
+        double voiceMarkMaxLow = result1.getMeasurement();
+        double voiceMarkMaxHigh = voiceMarkMaxLow;
+
+
+        int highCpu = (3 * numCPUs) / 4;
```

Comment:
`int lowCpu = (1 * numCPUs) / 4;
int highCpu = (3 * numCPUs) / 4;`

This way of choosing the CPU ids is tricky.
- We assume that the CPU numbering depends on their capacity.
- We assume that both lowCpu and highCpu are online.
- Again, we assume the existence of 2 only architectures.

With these assumptions, wouldn't it be similar to
`int lowCpu = 0;
int highCpu = numCPUs - 1;`
?

Do you want to avoid the leftmost and rightmost ids?

Reply:
We should run on every CPU. I was just lazy and didn't want to wait that long.
I will add a TODO for a future version.

Reply:
And yes I was avoiding the zeroth CPU because I think it is often busy.